Malloc should fail on too large allocations

xbauch opened the pull request:
via goto-checker with a dedicated cmdl option and a property that acknowledges pointer-width and object-size.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented:
This may fix #4604.

xbauch commented:
It seems that we have a few regression tests that call `malloc` with nondeterministic value (of type `size_t`) without assuming reasonable upper bound.

chrisr-diffblue commented on src/ansi-c/library/stdlib.c:
```diff
@@ -115,6 +115,9 @@ inline void *malloc(__CPROVER_size_t malloc_size)
   // and __CPROVER_allocate doesn't, but no one cares
   __CPROVER_HIDE:;
+  // 8 is the default object bits for ANSI-C and C++
+  if (malloc_size >= ((__CPROVER_size_t)1<<(64-8)))
```
Can't the object bits size be changed on the command line with `--object-bits` ? and if so, this probably needs to accommodate that. Ditto for the hardcoded `64`, which would probably be wrong if `--32` is specified to cbmc :-)

xbauch replied:
I moved the check into `goto-checker`, i.e. the `stdlib.c` implemetation is unchanged, we can use the pointer-width and object-bits from config, and the whole thing is behind an option, so it won't affect the existing tests.

chrisr-diffblue commented on regression/cbmc/malloc-too-large/main.c:
```diff
@@ -0,0 +1,11 @@
+
+int main()
+{
+  int *p = malloc(SIZE_MAX);
```
It would also be nice to have a test for a size that is 1 byte larger than the maximum CBMC can support, and also for a size that is exactly the same as the largest CBMC can support - just to test the boundary condition.

xbauch replied:
Added.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 640cdccbe1b06545e7cbb7f5b615c1f39505ee81).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143393509

chrisr-diffblue commented on regression/cbmc/malloc-too-large-almost/main.c:
```diff
@@ -0,0 +1,13 @@
+{
+  size_t pointer_width = sizeof(void *) * 8;
+  size_t object_bits = 8; // by default
+  size_t cbmc_max_alloc_size = (size_t)1 << (pointer_width - object_bits);
```
A really tiny nit pick, so feel free to ignore me :-) But I'd personally write this as:
```
size_t cbmc_max_alloc_size = ((size_t)1 << (pointer_width - object_bits))-1;
int *p2 = malloc(cbmc_max_alloc_size); // try to allocate exactly the max
```
just so that the variable names really do mean what they say. Ditto for the other test case - but like I say, it's a really minor nit for me to pick...

chrisr-diffblue commented on src/analyses/goto_check.h:
```diff
@@ -55,6 +56,7 @@ void goto_check(
   " --nan-check                  check floating-point for NaN\n" \
   " --no-built-in-assertions     ignore assertions in built-in library\n" \
   " --enum-range-check           checks that all enum type expressions have values in the enum range\n" /* NOLINT(whitespace/line_length) */ \
+  " --malloc-max-check           check that CBMC can represent the allocated object of requested size" /* NOLINT(whitespace/line_length) */ \
```
Is a better name for this option "malloc-size-check" ? Don't have a strong opinion.

chrisr-diffblue reviewed (approved):
Generally looks good to me, and I like that it's moved into goto-check. Apologies for the niggle comments - take them as suggestions, and not absolute requirements.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b73f9689ebadceadb4fe1c4d59a963b5a2b4ef9d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143440844

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -1941,6 +1943,30 @@ void goto_checkt::goto_check(
+                         minus_exprt{pointer_bits, object_bits}};
+          add_guarded_property(
+            binary_relation_exprt{rhs_side_effect.op0(), ID_lt, max_alloc_size},
+            "malloc_size < 2 ^ (pointer_width - object_size)",
```
`object_size` -> `object_id_width`?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0c2c362bb817a7c47212b49c726b39ffdac02767).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143549138

martin-cs commented:
On Tue, 2020-01-07 at 06:40 -0800, Chris Ryder wrote:
> Generally looks good to me, and I like that it's moved into goto-
> check.

I'm not sure I do.  Keeping the modelling of malloc in one place, in
the library models seems, to me, to be important.  Setting CPROVER
globals to control max limits, if malloc can fail, etc. seems a more
flexible option and keeps all the modelling together.

danielsn commented:
> It seems that we have a few regression tests that call `malloc` with nondeterministic value (of type `size_t`) without assuming reasonable upper bound.

I see two ways to handle `malloc(size)` where `size` is too large:
1. `assert(size <= MAX_MALLOC_SIZE)`
2. `if(size > MAX_MALLOC_SIZE) return NULL`

hannes-steffenhagen-diffblue commented on regression/goto-instrument/insert-final-assert-false1/main.c:
```diff
@@ -3,6 +3,12 @@
+  size_t pointer_width = sizeof(void *) * 8;
+  size_t object_bits = 8; // by default
+  size_t cbmc_max_alloc_size = ((size_t)1 << (pointer_width - object_bits)) - 1;
+  __CPROVER_assume(n <= cbmc_max_alloc_size);
```
May be worth adding this as an exposed constant `__CPROVER_max_alloc_size`? Seems to me this might come up fairly regularly in user scenarios as well, and it’d probably be a bit of a PITA to keep redefining this.

hannes-steffenhagen-diffblue reviewed (approved):
LGTM, @martin-cs are you happy with this iteration (changes moved to `stdlib.c`)?

chrisr-diffblue commented:
@xbauch Did you see @danielsn comment? I wonder if its worth adding a bit more user control so that the end user could decided which model (assert, versus return null) they wish to use? I guess that could even potentially be exposed via another global, e.g. `__CPROVER_malloc_failure_mode` or something like that? Much like the rounding mode variable... I guess that having a global like that could even mean that a user could write a harness function that could even change the malloc failure mode mid-harness...

xbauch commented:
@chrisr-diffblue I did see it and we experimented with `malloc` returning `NULL`, but the rest of CBMC does not like that.

danpoe commented on src/ansi-c/library/stdlib.c:
```diff
@@ -115,6 +115,15 @@ inline void *malloc(__CPROVER_size_t malloc_size)
+    "max allocation size exceeded");
+  __CPROVER_assume(
+    malloc_size < ((__CPROVER_size_t)1
+                   << (__CPROVER_pointer_width - __CPROVER_object_bits)));
```
Given that object offsets are signed (see #5096), I think this should be `malloc_size <= ((__CPROVER_size_t)1 << (__CPROVER_pointer_width - __CPROVER_object_bits - 1))`. Also we can use `<=` as the maximum allocation size is one larger than the largest offset we can represent.

xbauch replied:
Nice catch, thanks.

xbauch commented:
@chrisr-diffblue What do you think now? The first commit (19fcd57f385e8740b9ad84be9e2f8236c15cdf72) now contains two options to choose how malloc should fail.

martin-cs commented on regression/cbmc/array_constraints1/test.desc:
```diff
@@ -4,6 +4,6 @@ main.c
 ^SIGNAL=0$
 ^VERIFICATION FAILED$
-^\*\* 2 of 14
+^\*\* 2 of \d+
```
Why `\d+` ? shouldn't the number of VCs be known and static?

chrisr-diffblue replied:
Agree with Martin here - it's probably good that the number of VC's should be checked in the tests. I appreciate that is an added maintenance issue but we wouldn't want to silently regress the number of VCs either

martin-cs commented on regression/cbmc/pointer-overflow1/test.desc:
```diff
@@ -5,7 +5,7 @@ main.c
 ^\[main\.overflow\.\d+\] line \d+ (pointer )?arithmetic overflow on .*sizeof\(signed int\) .* : SUCCESS
 ^VERIFICATION FAILED$
-^\*\* 8 of 11 failed
+^\*\* 8 of \d+ failed
```
Same again.

martin-cs commented on regression/cbmc/r_w_ok1/test.desc:
```diff
@@ -2,7 +2,7 @@ CORE
 
 __CPROVER_[rw]_ok\(arbitrary_size, n \+ 1\): FAILURE$
-^\*\* 2 of 10 failed
+^\*\* 2 of \d+ failed
```
Another one.

martin-cs commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1041,6 +1041,9 @@ void cbmc_parse_optionst::help()
     "Backend options:\n"
     " --object-bits n              number of bits used for object addresses\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --malloc-fail-assert         set malloc failure mode to assert-then-assume\n"
```
I think these should really be in `Program instrumentation options` because they are controlling what properties are checked.

chrisr-diffblue replied:
Agree with Martin here too.

martin-cs commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -155,6 +155,11 @@ void ansi_c_internal_additions(std::string &code)
+      std::to_string(config.ansi_c.malloc_failure_mode)+";\n"
+    CPROVER_PREFIX "size_t " CPROVER_PREFIX "max_malloc_size="+
+    std::to_string(1 << (config.ansi_c.pointer_width -
+                         config.bv_encoding.object_bits - 1))+";\n"
```
Yes! Nice!

martin-cs reviewed (approved):
Thank you for reworking this.  I think this is a much simpler, compact and flexible way of getting this functionality.

chrisr-diffblue commented on src/ansi-c/library/stdlib.c:
```diff
@@ -115,23 +115,43 @@ inline void *malloc(__CPROVER_size_t malloc_size)
 
-  // make sure it's not recorded as deallocated
-  __CPROVER_deallocated=(malloc_res==__CPROVER_deallocated)?0:__CPROVER_deallocated;
+    if(__CPROVER_malloc_failure_mode == 1)
```
In an ideal world :-) rather than magic numbers here for the failure modes, these would be static named constants. Bonus points of those named constants were exposed to user programs as __CPROVER_blah constants.

xbauch replied:
Done.

chrisr-diffblue commented on src/util/config.cpp:
```diff
@@ -1093,6 +1093,16 @@ bool configt::set(const cmdlinet &cmdline)
+      "at most one malloc failure mode is acceptable", "--malloc-fail-null"};
+  }
+  if(cmdline.isset("malloc-fail-null"))
+    ansi_c.malloc_failure_mode = 1;
```
Again, would be nice to avoid the magic numbers here.

chrisr-diffblue reviewed (changes requested):
Looks really good - two main things I'd like to see before merging:
- Don't replace the VC numbers with patterns in regression tests
- Don't use magic numbers for malloc_failure_mode

chrisr-diffblue commented:
Oh, and one final thing that would make me really happy :-) is if the user docs were updated to describe this new feature....

xbauch commented:
@chrisr-diffblue 

> Looks really good - two main things I'd like to see before merging:
> 
> * Don't replace the VC numbers with patterns in regression tests

Went to exact numbers

> * Don't use magic numbers for malloc_failure_mode

There are now constants in `cprover.h` and an enum in `config.ansi_c`.
Also added some cprover-manual docu.

It's in a separate commit (https://github.com/diffblue/cbmc/commit/ed741ecd7b8353a22d46193e8cb8dbacb5e352af) for reviewing -- will be squashed before merging (if approved).

chrisr-diffblue reviewed (approved):
That looks great - many thanks for fixing all the nitpicks!

xbauch commented:
test group | before | after | slowdown
-- | -- | -- | --
cbmc-CORE | 127.9 | 146.88 | 1.14839718530102
cbmc-paths-lifo-CORE | 159.87 | 182.46 | 1.14130230812535
cbmc-library-CORE | 19.11 | 22.46 | 1.1753008895866
goto-analyzer-CORE | 15.89 | 17.75 | 1.11705475141598
ansi-c-CORE | 70.67 | 72.89 | 1.03141361256545
cbmc-cprover-smt2-CORE | 143.8 | 159.32 | 1.10792767732962
goto-instrument-CORE | 67.93 | 69.95 | 1.02973649344914
jbmc-CORE | 179.93 | 196.51 | 1.09214694603457
strings-smoke-tests-CORE | 68.95 | 69.65 | 1.01015228426396
strings-smoke-tests-symex-driven-lazy-loading-CORE | 260.05 | 300.16 | 1.15423956931359
jbmc-symex-driven-lazy-loading-CORE | 725.48 | 802.43 | 1.1060677068975
jbmc-strings-CORE | 317.44 | 327.24 | 1.03087197580645
jbmc-strings-symex-driven-lazy-loading-CORE | 950.72 | 1018.15 | 1.07092519353753
total | 2053.73 | 2233.66 | 1.0876113218388

There is some time regression but the CI is suddenly takes >1800s to compile (before it was ~50s).

smowton commented:
That's ccache.

xbauch commented:
> That's ccache.

Do you mean that it's ccache that's timing out or that the difference between build times is caused by ccache?

chrisr-diffblue commented:
Difference in build times, I suspect. Because you've touched a few header files that are reasonably widely included, ccache is going to end up thinking it does actually need to rebuild a lot of things, rather than just pulling them from the cache.

chrisr-diffblue commented:
Since the failing Travis job did actually get as far as saving it's ccache, there's a reasonably high (by Travis standards...) that simply re-running the failed job will make it pass.

xbauch commented:
> Since the failing Travis job did actually get as far as saving it's ccache, there's a reasonably high (by Travis standards...) that simply re-running the failed job will make it pass.

This was my 4th attempt.

chrisr-diffblue commented:
:-s Well that sucks...

chrisr-diffblue commented:
Looking a little deeper... the *previous* Travis run had two of the OSX configs timeout... *one* of those configs got as far as saving its cache, the other didnt. This most recent Travis run has only a single Travis failure - the config that previously timed out before saving it's cache - while the config that did manage to save it's cache has now passed. So I think it is worth one more attempt to re-run the Travis jobs since I think all the configs have now saved updated caches.

petr-bauch committed:
Malloc should fail on too large allocations

1: two new options to choose between fail by assert-then-assume and return null
2: `__CPROVER_max_malloc_size` is introduced
3: some tests for boundary conditions
4: user-level documentation

petr-bauch committed:
Update the constant propagation tests

More assignments due to the `__CPROVER_malloc_max_size` and
`__CPROVER_malloc_failure_mode` instantiations.

petr-bauch committed:
Update the expected number of assertions

and test them to any number.

petr-bauch committed:
Remove undefined behaviour from pointer-byte-extract5

specifically the C90-style struct hack with non-flexible member.

xbauch commented:
The changes in this PR cause ~7% slow down (the CORE test suite from 1560s to 1674s). It might be avoided if we figured out how to introduce `#ifdef`s in the library models based on user options. Any ideas @peterschrammel? Is it worth the trouble?

petr-bauch committed:
Snapshot harness test fix

before we were extracting the value for the symbol to be used as array size, but
the initialization to the extracted value is not guaranteed to happen before the
initialization of the array. So we avoid extracting this value from the snapshot
altogether.
